File: pomade-signer/src/mailer/sendlayer.rs
use serde_json::json;

use crate::mailer::{Email, MailFuture, Mailer};

pub struct SendlayerMailer {
    pub api_key: String,
}

impl Mailer for SendlayerMailer {
    fn send(&self, from_email: &str, from_name: &str, email: Email) -> MailFuture {
        let key = self.api_key.clone();
        let from_email = from_email.to_string();
        let from_name = from_name.to_string();
        Box::pin(async move {
            let res = reqwest::Client::new()
                .post("https://console.sendlayer.com/api/v1/email")
                .bearer_auth(&key)
                .json(&json!({
                    "From": {"name": from_name, "email": from_email},
                    "To": [{"name": "", "email": email.to}],
                    "Subject": email.subject,
                    "ContentType": "HTML",
                    "HTMLContent": email.html,
                    "PlainContent": email.text,
                }))
                .send()
                .await
                .map_err(|e| e.to_string())?;

            if res.status().is_success() {
                Ok(())
            } else {
                Err(format!("sendlayer error: {}", res.status()))
            }
        })
    }
}

File: pomade-signer-go/mailer/sendlayer.go
package mailer

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

type SendlayerMailer struct {
	Client *http.Client
	APIKey string
}

func (m SendlayerMailer) Send(ctx context.Context, fromEmail string, fromName string, email Email) error {
	body, _ := json.Marshal(map[string]any{
		"from": map[string]any{
			"email": fromEmail,
			"name":  fromName,
		},
		"to":      []any{map[string]any{"email": email.To}},
		"subject": email.Subject,
		"text":    email.Text,
		"html":    email.HTML,
	})
	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, "https://console.sendlayer.com/api/v1/email", bytes.NewReader(body))
	req.Header.Set("Authorization", "Bearer "+m.APIKey)
	req.Header.Set("Content-Type", "application/json")
	res, err := m.Client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode > 299 {
		return fmt.Errorf("sendlayer error: %s", res.Status)
	}
	return nil
}
